Add Price Optimization Notebook

FengShi0705 opened the pull request:
*Description of changes:*
Adding price optimization notebook.

We prepare to publish an AWS blog post where we used Amazon Braket quantum annealer to solve a price optimization problem. We are adding this notebook to complement the blog post as an asset.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

christianbmadsen reviewed (commented):
Remove the reference to the blog post in the example notebook please.

FengShi0705 commented on examples/quantum_annealing/price_optimization/qubo_dynamic_pricing.py:
```diff
@@ -0,0 +1,255 @@
+    model = (-obj).compile().to_bqm()
+    num_shots = 10000
+
+    sampler = BraketDWaveSampler(('amazon-braket-481358cc730d', 'qubo'),
```
Thanks chris for pointing it out, these two item has now been addressed in the new commit

christianbmadsen reviewed (commented):
Cell 26 takes a long time to run and submits a substantial amount of tasks against DWave. I recommend that you comment out as default similar to cell 9 here:https://github.com/aws/amazon-braket-examples/blob/main/examples/braket_features/Using_the_tensor_network_simulator_TN1.ipynb

virajvchaudhari commented:
Hi @FengShi0705, Thanks for submitting the notebook. I just have some questions/comments hoping you can address. 
1. Can we handle the required imports in the top most cell and not repeat them in the below cells. You can refer other notebooks for this. 
2. Can you remove the comment in the last cell `#b=300`, if that's not required? 
3. Can we have a proper spacing after assigning the variable and the working code. You can refer to this notebook to understand how we are aligning the spacing. https://github.com/aws/amazon-braket-examples/blob/main/examples/quantum_annealing/Dwave_Factoring/Dwave_factoring.ipynb
4. As Christian pointed out the cell 26 takes a long time to run, may be we can comment out that cell and add a note that will allow users to understand that it will take long time to run and whether they want to execute it. You can refer this  notebook: https://github.com/aws/amazon-braket-examples/blob/main/examples/braket_features/Using_the_tensor_network_simulator_TN1.ipynb

virajvchaudhari commented on examples/quantum_annealing/price_optimization/qubo_dynamic_pricing.py:
```diff
@@ -0,0 +1,256 @@
+    return ((demand-vol_bounday-slack)**2)*Ld
+
+
+def main(
```
nit: Should we have a different function name instead of `main`?

FengShi0705 replied:
yes, now we change the name of `main` to `optimize`

FengShi0705 commented:
> Cell 26 takes a long time to run and submits a substantial amount of tasks against DWave. I recommend that you comment out as default similar to cell 9 here:https://github.com/aws/amazon-braket-examples/blob/main/examples/braket_features/Using_the_tensor_network_simulator_TN1.ipynb

done

FengShi0705 commented:
> import matplotlib.pyplot as plt
> 
> We are already importing this, can we remove these imports from every cell again?

done

FengShi0705 commented:
> Hi @FengShi0705, Thanks for submitting the notebook. I just have some questions/comments hoping you can address.
> 
> 1. Can we handle the required imports in the top most cell and not repeat them in the below cells. You can refer other notebooks for this.
> 2. Can you remove the comment in the last cell `#b=300`, if that's not required?
> 3. Can we have a proper spacing after assigning the variable and the working code. You can refer to this notebook to understand how we are aligning the spacing. https://github.com/aws/amazon-braket-examples/blob/main/examples/quantum_annealing/Dwave_Factoring/Dwave_factoring.ipynb
> 4. As Christian pointed out the cell 26 takes a long time to run, may be we can comment out that cell and add a note that will allow users to understand that it will take long time to run and whether they want to execute it. You can refer this  notebook: https://github.com/aws/amazon-braket-examples/blob/main/examples/braket_features/Using_the_tensor_network_simulator_TN1.ipynb
> 5. Can we also change some variable names instead of `a` `b`, that would give more context about the variables?

Good suggestions. We addressed point 1~4  as suggested
 
For point 5, could we keep the naming of a,b as this aims to be consistent with the math equation in the text?

FengShi0705 commented:
<p>Done</p>

---

 <a href='https://app.reviewnb.com/aws/amazon-braket-examples/pull/88/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

FengShi0705 commented:
<p>Done</p>

---

 <a href='https://app.reviewnb.com/aws/amazon-braket-examples/pull/88/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

FengShi0705 commented:
<p>Done, complete it now</p>

---

 <a href='https://app.reviewnb.com/aws/amazon-braket-examples/pull/88/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

FengShi0705 commented:
<p>Done</p>

---

 <a href='https://app.reviewnb.com/aws/amazon-braket-examples/pull/88/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

FengShi0705 commented:
<p>Done</p>

---

 <a href='https://app.reviewnb.com/aws/amazon-braket-examples/pull/88/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

FengShi0705 commented:
<p>Done</p>

---

 <a href='https://app.reviewnb.com/aws/amazon-braket-examples/pull/88/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

FengShi0705 commented:
<p>Done</p>

---

 <a href='https://app.reviewnb.com/aws/amazon-braket-examples/pull/88/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

FengShi0705 commented:
<p>yes, done, reduced the size of output printing</p>

---

 <a href='https://app.reviewnb.com/aws/amazon-braket-examples/pull/88/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

FengShi0705 commented:
<p>Done.</p><p>Good point, the rounding is a legacy procedure for previous version. There is no need for rounding to take place, so the wording around "rounding" is now removed .</p>

---

 <a href='https://app.reviewnb.com/aws/amazon-braket-examples/pull/88/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

FengShi0705 commented:
<p>Done</p>

---

 <a href='https://app.reviewnb.com/aws/amazon-braket-examples/pull/88/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

FengShi0705 commented:
<p>Done</p>

---

 <a href='https://app.reviewnb.com/aws/amazon-braket-examples/pull/88/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

FengShi0705 commented:
<p>Done.</p><p> </p><p>Good point, A sentence is now added to explain the discrete price choices. <code>According to certain business requirements, the possible prices are usually defined to be a set of discrete price values (e.g.&nbsp;$5.99,&nbsp;$9.99,&nbsp;$14.99).</code></p>

---

 <a href='https://app.reviewnb.com/aws/amazon-braket-examples/pull/88/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

FengShi0705 commented:
<p>Done.</p><p> </p><p>Now added comments in the code: <code># here, we use a list to represent the sequence of prices from day T-n to day T+n-1.</code></p>

---

 <a href='https://app.reviewnb.com/aws/amazon-braket-examples/pull/88/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

FengShi0705 commented:
<p>Done</p>

---

 <a href='https://app.reviewnb.com/aws/amazon-braket-examples/pull/88/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

FengShi0705 commented:
<p>Done. comment added</p>

---

 <a href='https://app.reviewnb.com/aws/amazon-braket-examples/pull/88/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

FengShi0705 commented:
<p>Done as suggested</p>

---

 <a href='https://app.reviewnb.com/aws/amazon-braket-examples/pull/88/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

FengShi0705 commented:
<p>Done. and a reference to the formula is added</p>

---

 <a href='https://app.reviewnb.com/aws/amazon-braket-examples/pull/88/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

FengShi0705 commented:
<p>Done</p>

---

 <a href='https://app.reviewnb.com/aws/amazon-braket-examples/pull/88/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

FengShi0705 commented:
<p>Perfect suggestion. Done</p>

---

 <a href='https://app.reviewnb.com/aws/amazon-braket-examples/pull/88/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

FengShi0705 commented:
<p>Done. reduced the printing size. The rows and columns are now explained in the text. <code>Each row represents an optimized solution. The columns starting with&nbsp;X_&nbsp;shows the value of the binary decision variables&nbsp;$x_{t,k}$&nbsp;and the&nbsp;<code>energy</code>&nbsp;column shows the negative value of objective, i.e.&nbsp;−𝐻</code></p>

---

 <a href='https://app.reviewnb.com/aws/amazon-braket-examples/pull/88/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

FengShi0705 commented:
<p>Done</p>

---

 <a href='https://app.reviewnb.com/aws/amazon-braket-examples/pull/88/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

FengShi0705 commented:
<p>Done</p>

---

 <a href='https://app.reviewnb.com/aws/amazon-braket-examples/pull/88/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

FengShi0705 commented:
<p>Done</p>

---

 <a href='https://app.reviewnb.com/aws/amazon-braket-examples/pull/88/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

FengShi0705 commented:
<p>Really great point. We missed the covariance in this formula.</p><p>Now we made following changes:</p><ol><li>we added covariance in this formula</li><li>in the code, we now added a function to calculate the covariance</li></ol>

---

 <a href='https://app.reviewnb.com/aws/amazon-braket-examples/pull/88/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

FengShi0705 commented:
<p>Done. Added explanation sentence <code>This small but non-zero value is due to floating point precision.</code></p>

---

 <a href='https://app.reviewnb.com/aws/amazon-braket-examples/pull/88/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

FengShi0705 commented:
<p>Done</p>

---

 <a href='https://app.reviewnb.com/aws/amazon-braket-examples/pull/88/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

FengShi0705 commented:
<p>Done</p>

---

 <a href='https://app.reviewnb.com/aws/amazon-braket-examples/pull/88/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

FengShi0705 commented:
<p>Done, added wikipedian link</p>

---

 <a href='https://app.reviewnb.com/aws/amazon-braket-examples/pull/88/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

FengShi0705 commented on examples/quantum_annealing/price_optimization/QUBO-Pricing.ipynb:
```diff
@@ -0,0 +1,2839 @@
+{
```
Thanks for the comments.
Done: `We start by importing Amazon braket SDK and Dwave SDK to be used for quantum annealing in this use case. In addition, we import sklearn to fit the demand estimation model.`

FengShi0705 commented on examples/quantum_annealing/price_optimization/QUBO-Pricing.ipynb:
```diff
@@ -0,0 +1,2839 @@
+{
```
Done
`Objective with expected revenue and prediction variance:`

FengShi0705 commented on examples/quantum_annealing/price_optimization/QUBO-Pricing.ipynb:
```diff
@@ -0,0 +1,2839 @@
+{
```
Done
`Objective with additional equality constraints:`

JordanAWS commented on examples/quantum_annealing/price_optimization/QUBO-Pricing.ipynb:
`s3_folder` no longer needed. Can remove line.

JordanAWS commented on examples/quantum_annealing/price_optimization/QUBO-Pricing.ipynb:
```diff
@@ -0,0 +1,2855 @@
+    "## run dwave quantum annealing\n",
+    "num_shots = 10000\n",
+    "\n",
+    "sampler = BraketDWaveSampler(s3_folder,'arn:aws:braket:::device/qpu/d-wave/Advantage_system4')\n",
```
`s3_folder` no longer required. Can be changed to
 `sampler = BraketDWaveSampler(device_arn='arn:aws:braket:::device/qpu/d-wave/Advantage_system4')`

JordanAWS commented on examples/quantum_annealing/price_optimization/QUBO-Pricing.ipynb:
```diff
@@ -0,0 +1,2855 @@
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Enter the S3 bucket you created during onboarding in the code below\n",
```
Braket now supports default s3 buckets, so this step is no longer required. See other changes below reflecting this. I pushed some edits to your script to reflect this as well.

JordanAWS commented on examples/quantum_annealing/price_optimization/QUBO-Pricing.ipynb:
```diff
@@ -0,0 +1,2855 @@
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
```
May want to say here that this cell takes a few minutes to run.

JordanAWS commented on examples/quantum_annealing/price_optimization/QUBO-Pricing.ipynb:
The whitespace seems to have gotten messed up somehow. Please paste in this, which should work instantly with the standard Command/CTRL -> / for commenting/uncommenting multi lines. 
```
# beta_options=[1e2,1e3,1e4,1e5,1e6,1e7,1e8,1e9]
# results = {beta_option:[] for beta_option in beta_options}
# beta_samples = beta_options*6

# for beta_i in beta_samples:
#     max_revenue, prediction_variance, energy, opt_demand, opt_prices, rev_std = optimize(
#        a = a,
#        b = b,
#        data_x = data_x,
#        selected_hist_prices = p_data,
#        price_levels = price_levels,
#        Lp = Lp,
#        Ld = Ld,
#        sigma = sigma ,
#        beta=beta_i,
#        vol_bound=None,
#    )
#     print('---')
#     print(f"beta_sample:{beta_i}")
#     print(f"max_revenue:{max_revenue}")
#     print(f"prediction_variance:{prediction_variance}")
#     print(f"energy:{energy}")
#     print(f"opt_demand:{opt_demand}")
#     print(f"opt_prices:{opt_prices}")
#     print(f"rev_std: {rev_std}")
#     print(f"penalty_term: {max_revenue-beta_i*prediction_variance+energy}")
#     results[beta_i].append([energy, max_revenue, prediction_variance, rev_std])
```

JordanAWS commented on examples/quantum_annealing/price_optimization/qubo_dynamic_pricing.py:
```diff
@@ -0,0 +1,282 @@
+    :return: variance
+    """
+    n_samples, t = data_x.shape
+    ones = np.ones((n_samples, 1), dtype=np.float)
```
Change to builtin `float`

JordanAWS commented on examples/quantum_annealing/price_optimization/qubo_dynamic_pricing.py:
```diff
@@ -0,0 +1,282 @@
+    :return: variance
+    """
+    n_samples, t = data_x.shape
+    ones = np.ones((n_samples, 1), dtype=np.float)
```
Change to built-in `float`

JordanAWS commented on examples/quantum_annealing/price_optimization/QUBO-Pricing.ipynb:
Change to built-in `float`

JordanAWS reviewed (changes requested):
Thank you so much for taking the time to create this demo notebook for Braket! Fantastic description of the use case. I've added a number of edit requests below, mostly style things, and also some code updates now that we have the option to do default s3 buckets.

dtrenev commented:
Thank you for your contribution to this project. We're closing this PR as part of repository maintenance, since it has been inactive for over 3 years and the codebase has evolved significantly since then. We appreciate your continued interest.
